pkg/exporter/syslogexporter/syslog.go

```diff
@@ -0,0 +1,206 @@
+const versionRFC5424 = 1
+
+const formatRFC5424Str = "RFC5424"
+const formatRFC3164Str = "RFC3164"
+
+const priority = "priority"
+const facility = "facility"
+const version = "version"
+const timestamp = "timestamp"
+const hostname = "hostname"
+const app = "appname"
+const pid = "proc_id"
+const msgId = "msg_id"
+const structuredData = "structured_data"
+const message = "message"
+
+type Syslog struct {
+	hostname                 string
+	network                  string
+	addr                     string
+	format                   string
+	app                      string
+	pid                      int
+	additionalStructuredData []string
+	tlsConfig                *tls.Config
+	logger                   *zap.Logger
+	mu                       sync.Mutex
+	conn                     net.Conn
+}
+
+func Connect(logger *zap.Logger, cfg *Config, tlsConfig *tls.Config, hostname string, pid int, app string) (*Syslog, error) {
+	s := &Syslog{
+		logger:                   logger,
+		hostname:                 hostname,
+		network:                  cfg.Protocol,
+		addr:                     fmt.Sprintf("%s:%d", cfg.Endpoint, cfg.Port),
+		format:                   cfg.Format,
+		tlsConfig:                tlsConfig,
+		pid:                      pid,
+		app:                      app,
+		additionalStructuredData: cfg.AdditionalStructuredData,
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := s.connect()
+	if err != nil {
+		return nil, err
+	}
+	return s, err
+}
+
+func (s *Syslog) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.conn != nil {
+		err := s.conn.Close()
+		s.conn = nil
+		return err
+	}
+	return nil
+}
+
+func (s *Syslog) connect() error {
+	if s.conn != nil {
+		s.conn.Close()
+		s.conn = nil
+	}
+	var err error
+	if s.tlsConfig != nil {
+		s.conn, err = tls.Dial("tcp", s.addr, s.tlsConfig)
+	} else {
+		s.conn, err = net.Dial(s.network, s.addr)
+	}
+	return err
+}
+
+func (s *Syslog) Write(msg map[string]any) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	/* isFormatCorrect := s.validateFormat(msg)
+
+	if !isFormatCorrect && s.dropInvalidMsg {
+		s.logger.Debug("Invalid message format",
+			zap.String("format", s.format),
+			zap.String("msg", msg),
+		)
+		return nil
+	} else if !isFormatCorrect {
+		msg = s.formatMsg(msg)
+	}*/
+	s.addStructuredData(msg)
+	msgStr := s.formatMsg(msg)
+
+	if s.conn != nil {
+		if err := s.write(msgStr); err == nil {
+			return nil
+		}
+	}
+	if err := s.connect(); err != nil {
+		return err
+	}
+
+	return s.write(msgStr)
+}
+
+func (s *Syslog) write(msg string) error {
+	// check if logs contains new line character at the end, if not add it
+	if !strings.HasSuffix(msg, "\n") {
+		msg = fmt.Sprintf("%s%s", msg, "\n")
+	}
+	_, err := fmt.Fprint(s.conn, msg)
+	return err
+}
+
+func (s *Syslog) formatMsg(msg map[string]any) string {
+	switch s.format {
+	case formatRFC3164Str:
+		return formatRFC3164(msg)
+	case formatRFC5424Str:
+		return formatRFC5424(msg)
+	default:
+		return ""
```

Comment:
we should panic here. This situation shouldn't happen at all

Reply:
done